cloudformation/template.yml

```diff
@@ -0,0 +1,339 @@
+          - DesiredVCPUs
+          - MaximumVCPUs
+      -
+        Label:
+          default: Docker Container Configuration
+        Parameters:
+          - RepositoryName
+          - ImageName
+          - ImageTag
+          - InstanceVCPUs
+          - InstanceMemory
+      -
+        Label:
+          default: Batch Configuration
+        Parameters:
+          - JobDefinitionName
+          - BatchServiceRolePolicyARN
+          - SpotFleetServiceRolePolicyARN
+    ParameterLabels:
+      Project:
+        default: Name
+      Environment:
+        default: Environment
+      Region:
+        default: AWS Region
+      AMI:
+        default: AMI
+      KeyName:
+        default: SSH Key Name
+      SpotFleetBidPercentage:
+        default: Spot Fleet Bid Percentage
+      MinimumVCPUs:
+        default: Minimum vCPU Count
+      DesiredVCPUs:
+        default: Desired vCPU Count
+      MaximumVCPUs:
+        default: Maximum vCPU Count
+      InstanceTypes:
+        default: Instance Types
+      JobDefinitionName:
+        default: Job Definition Name
+      InstanceVCPUs:
+        default: vCPU Limit
+      InstanceMemory:
+        default: Memory Limit
+      BatchServiceRolePolicyARN:
+        default: Batch Service Role
+      SpotFleetServiceRolePolicyARN:
+        default: Spot Fleet Service Role
+      RepositoryName:
+        default: Repository Name
+      ImageName:
+        default: Image Name
+      ImageTag:
+        default: Image Tag
+      VPC:
+        default: VPC
+      SubnetIds:
+        default: Subnets
+
+Parameters:
+  Project:
+    Type: String
+    Default: Raster Vision CloudFormation
+    Description: Name of the project associated with the Batch job
+
+  Environment:
+    Type: String
+    Default: Testing
+    Description: Environment of the project being deployed (e.g. Staging, Development, Production)
+
+  Region:
+    Type: String
+    Default: 'us-east-1'
+    Description: AWS Region in which to deploy the Batch job
+
+  AMI:
+    Type: String
+    Default: 'ami-06681f03ed7f21643'
+    Description: Amazon Machine Image to use for the compute environment
+
+  KeyName:
+    Type: 'AWS::EC2::KeyPair::KeyName'
+    Description: An Amazon EC2 key pair name to use for SSH access to the cluster
+
+  SpotFleetBidPercentage:
+    Type: Number
+    Default: 100
+    Description: >
+      Minimum percentage that a Spot Instance price must be when compared with
+      the On-Demand price for that instance type before instances are launched
+
+  MinimumVCPUs:
+    Type: Number
+    Default: 0
+    Description: The minimum number of EC2 vCPUs that an environment should maintain
+
+  DesiredVCPUs:
+    Type: Number
+    Default: 0
+    Description: The desired number of EC2 vCPUS in the compute environment
+
+  MaximumVCPUs:
+    Type: Number
+    Default: 80
+    Description: The maximum number of EC2 vCPUs that an environment can reach
+
+  InstanceTypes:
+    Type: List<String>
+    Default: p2.xlarge
```

Comment:
It might be useful to add an `AllowedValues` attribute here for only the GPU instance types we want to target.

```yaml
AllowedValues:
  - ...
  - ...
```

Reply:
Nice.

Reply:
Having some trouble getting `AllowedValues` to work properly with the `List<String>` type here. When `AllowedValues` is specified, CloudFormation always throws the following template syntax error, no matter the default value passed in:

> Template validation error: Each value of parameter 'InstanceTypes' must be one of AllowedValues

Here's the relevant block:

```yaml
  InstanceTypes:
    Type: List<String>
    Default: p3.xlarge
    AllowedValues:
      - p2.xlarge
      - p2.8xlarge
      - p2.16xlarge
```

No docs on this online as far as I can tell, either -- every example I could find uses `AllowedValues` for a `String` or `Number` type, which makes me suspect that it's not supported for `List`s. Are you comfortable punting on this until we get a sense of how useful it would be?

Reply:
This was the snippet I was referencing in standup:

https://github.com/raster-foundry/prediction-api/blob/a7a69abb29ce9f9c33e27ec1e9dbc8a6c732e047/deployment/cloudformation/hot-prediction-endpoint-template.yaml#L17-L27

I think the point you made still stands though. This example is not of type list.